datafusion/physical-plan/src/sorts/builder.rs

```diff
@@ -21,17 +21,15 @@ use arrow::record_batch::RecordBatch;
 use datafusion_common::Result;
 use datafusion_execution::memory_pool::MemoryReservation;
 
-#[derive(Debug, Copy, Clone, Default)]
-struct BatchCursor {
-    /// The index into BatchBuilder::batches
-    batch_idx: usize,
-    /// The row index within the given batch
-    row_idx: usize,
-}
+use super::batches::{BatchCursor, BatchId};
+use super::cursor::{Cursor, CursorValues};
+
+type YieldedSortOrder<C> = (Vec<SortOrder>, Vec<BatchCursor<C>>);
+type SortOrder = (BatchId, usize); // (batch_id, row_idx)
 
 /// Provides an API to incrementally build a [`RecordBatch`] from partitioned [`RecordBatch`]
 #[derive(Debug)]
-pub struct BatchBuilder {
+pub struct SortOrderBuilder<C: CursorValues> {
```

Comment:
Is the eventual plan to remove `batches` from this? It seems a little peculiar as a construction as it stands

Reply:
Yes, exactly. I just pushed [a commit with documentation](https://github.com/apache/arrow-datafusion/pull/7842/commits/222c12a4ee9ad0c847b66dfefc168a7aabfa9e53) -- to show where this was going.

I was trying to break up the change into 2 parts (see PR description). That decision seems to have created confusion. 😅

Reply:
I wonder if this should instead be called CursorInterleave or something, as I believe that is the operation it is actually performing?

Reply:
> That decision seems to have created confusion

Yeah perhaps you could roll that into this PR, whilst it will be larger, I think it will be easier to review